Product Form: Support viewing product page for coming soon sites

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #14924 
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
In an upcoming change, WooCommerce plugin will set all new sites to [coming soon mode](https://developer.woocommerce.com/?p=8762487&preview=true&_thumbnail_id=8762342) by default.

This PR updates the web view used in the product form for viewing product page to support viewing products in non-published sites.

Also: updated the background color of the product status badge in the product form.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
- Log in to a WPCom store that's set to coming soon with existing products.
- Navigate to the Products tab and select a published product.
- Select the ellipsis menu on the top right and tap View Product in Store.
- Confirm that you can see the product page with product details.
- Switch to a self-hosted site while still authenticated with WPCom.
- Confirm that viewing product page still work for this site.
- Log out and back in to a self-hosted site using site credentials.
- Confirm that viewing product page still work for this site.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->
Tested on simulator iPhone 16 Pro iOS 18.2 and confirmed that I can see the product page for all products in both public and coming-soon stores.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->
/ | Before | After 
--- | --- | ---
Coming soon WPCom store | <image src="https://github.com/user-attachments/assets/4aecda3b-329f-4479-91f0-96912eaa955d" width=320 /> | <image src="https://github.com/user-attachments/assets/0233edbc-845c-47fa-b004-1d7662685867" width=320 />



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

itsmeichigo committed:
Use authenticated web view to display product page

itsmeichigo committed:
Enable viewing product page for private products

itsmeichigo committed:
Fix navigation item for the product page view

itsmeichigo committed:
Update LabeledTextViewTableViewCell.swift

itsmeichigo committed:
Update unit test for editing products with non published status

itsmeichigo committed:
Update release notes

pmusolino reviewed (commented):
Something strange is happening in my testing store with this PR. 

Both with "coming soon" mode enabled or disabled. When opening the web page of every type of product, a Jetpack stats page is loading for me.
While, if I open a published product using `trunk`, everything works as expected.
It's like in the version with `AuthenticatedWebViewController`, the redirect doesn't work properly.


Screenshot of what I see:
![Simulator Screenshot - iPhone 16 Pro - 2025-01-21 at 17 58 36](https://github.com/user-attachments/assets/f6a31287-d2df-4b75-a5a8-f9c58edaeac1)

itsmeichigo commented:
Hey @pmusolino - can you confirm how you are authenticated in the app? E.g. with WPCom or application password?

Can you also try logging out and in again to see if you can reproduce the issue consistently? And does the issue happen to all your test sites?

pmusolino commented:
> with WPCom or application password?

I'm authenticated with WPCom

> Can you also try logging out and in again to see if you can reproduce the issue consistently? And does the issue happen to all your test sites?

- It happens just on my testing store. If I try to use your testing store or other testing stores, everything works properly.
- If I logout, then i login, I'm still able to reproduce the issue.

itsmeichigo committed:
Revert "Update unit test for editing products with non published status"

This reverts commit 5e30b40543b5a4ff6fe41e44b9247842427ace45.

itsmeichigo committed:
Revert "Enable viewing product page for private products"

This reverts commit ab372902f05bd8a605b6cf851e4a2cdb86517408.

itsmeichigo commented:
Thanks @pmusolino for the review! It looks like the authenticated web view doesn't work if you're authenticated with a WPCom account and open the product page on a self-hosted site. I'll convert this PR back to draft and see how this should be updated.

itsmeichigo committed:
Only authenticate preview page if possible

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/235">21.5</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo committed:
Check for Jetpack plugin before authenticate user

itsmeichigo committed:
Revert "Check for Jetpack plugin before authenticate user"

This reverts commit 26eef5c605f9353c7479bc3b4d9860ff7032b0b0.

itsmeichigo committed:
Update release notes

itsmeichigo commented:
Hi @pmusolino - I updated the PR with two changes:
- Disable preview for non-public products.
- Authenticate the preview page only for:
  - WPCom sites
  - Self-hosted sites when the user is authenticated with application password/nonce.

I updated the PR description accordingly - this PR is ready for another look!

itsmeichigo committed:
Merge branch 'trunk' into enhancement/14924-previewing-product-page

pmusolino commented on RELEASE-NOTES.txt:
```diff
@@ -7,6 +7,7 @@
 - [*] Receipts: Email receipts can now be sent to customers after failed payments using WooCommerce Stripe 9.1.0+. [https://github.com/woocommerce/woocommerce-ios/pull/14864].
 - [*] Order Creation: orders are always created using the store's currency, not the user's [https://github.com/woocommerce/woocommerce-ios/pull/14907]
 - [*] Dashboard: Updated the drag gesture on the Performance chart to show stats instantly. [https://github.com/woocommerce/woocommerce-ios/pull/14906]
+- [*] Product Form: Support previewing products on coming-soon stores. [https://github.com/woocommerce/woocommerce-ios/pull/14931]
```
nit: I wonder if for end users it's more clear "Product Detail" instead of "Product Form".

itsmeichigo replied:
We'll rewrite the message on the changelog anyway, I think this note is mostly for us developers. Since changing this will trigger CI again, I'll keep it as-is and merge the PR for the code freeze. I hope that's fine.

pmusolino reviewed (approved):
I tested all the cases mentioned in this PR, both with WPCom and Application Password login, and everything works fine now!

Just a nit on release-notes 👍
